File: 0547-number-of-provinces/0547-number-of-provinces.js
/**
 * @param {number[][]} isConnected
 * @return {number}
 */
class UF {
    constructor() {
        this.root = {}
        this.provinces = 0
    }
    insert(a) {
        if (!(a in this.root)) {
            this.root[a] = a
            this.provinces++
        }
    }
    find(a) {
        this.insert(a)
        if (this.root[a] !== a) {
            this.root[a] = this.find(this.root[a])
        }
        return this.root[a]
    }
    union(a, b) {
        const A = this.find(a)
        const B = this.find(b)
        if (A === B) return
        this.root[B] = A
        this.provinces--
    }
}

var findCircleNum = function(isConnected) {
    const uf = new UF()
    const n = isConnected.length
    for (let i=0; i < n; i++) {
        for (let j=0; j < n; j++) {
            if (isConnected[i][j] === 1) {
                uf.union(i, j)
            } else {
                uf.insert(i)
                uf.insert(j)
            }
        }
    }
    return uf.provinces
};

File: 0547-number-of-provinces/0547-number-of-provinces.py
class Solution:
    def bfs(self, isConnected: List[List[int]], city: int, visited: set) -> None:
        q = deque([city])
        while q:
            cur_city = q.popleft()
            if cur_city in visited:
                continue
            visited.add(cur_city)
            for next_city, yes in enumerate(isConnected[cur_city]):
                if yes == 1:
                    q.append(next_city)
    
    def dfs(self, isConnected: List[List[int]], city: int, visited: set) -> None:
        if city in visited:
            return
        visited.add(city)
        for next_city, yes in enumerate(isConnected[city]):
            if yes == 1:
                self.dfs(isConnected, next_city, visited)
    
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n, visited, provinces = len(isConnected), set(), 0
        for city in range(0, n):
            if city not in visited:
                provinces += 1
                self.bfs(isConnected, city, visited)
        return provinces
        
        # 
        
        n, visited, provinces = len(isConnected), set(), 0
        for city in range(0, n):
            if city not in visited:
                provinces += 1
                self.dfs(isConnected, city, visited)
        return provinces
        